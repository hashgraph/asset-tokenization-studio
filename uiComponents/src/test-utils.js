import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
