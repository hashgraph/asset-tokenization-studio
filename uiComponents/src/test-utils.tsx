import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";

const DefaultTheme = extendTheme(BasePlatformTheme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
