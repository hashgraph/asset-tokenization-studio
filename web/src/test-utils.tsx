import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
