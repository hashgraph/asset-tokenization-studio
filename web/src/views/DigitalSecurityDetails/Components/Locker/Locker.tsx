import { Center, HStack, Stack, useDisclosure, VStack } from "@chakra-ui/react";
import {
  Button,
  DefinitionList,
  Heading,
  PhosphorIcon,
  PopUp,
  SearchInputController,
  Text,
} from "io-bricks-ui";
import { useForm } from "react-hook-form";
import { isValidHederaId, required } from "../../../../utils/rules";
import { useRolesStore } from "../../../../store/rolesStore";
import { SecurityRole } from "../../../../utils/SecurityRole";
import { Link as RouterLink, useParams } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { RouterManager } from "../../../../router/RouterManager";
import { RouteName } from "../../../../router/RouteName";
import {
  GET_LOCKERS,
  useGetLockers,
} from "../../../../hooks/queries/useGetLockers";
import {
  GetLocksIdRequest,
  LockViewModel,
  ReleaseRequest,
} from "@hashgraph/asset-tokenization-sdk";
import { useState } from "react";
import { formatDate } from "../../../../utils/format";
import { DATE_TIME_FORMAT } from "../../../../utils/constants";
import { useRelease } from "../../../../hooks/mutations/useLocker";
import { useQueryClient } from "@tanstack/react-query";
import { WarningCircle } from "@phosphor-icons/react";

export const Locker = () => {
  const queryClient = useQueryClient();

  const { id = "" } = useParams();

  const { t } = useTranslation("security", {
    keyPrefix: "details.locker",
  });
  const { t: tButtons } = useTranslation("security", {
    keyPrefix: "details.locker.actions",
  });
  const { t: tRelease } = useTranslation("security", {
    keyPrefix: "details.locker.release",
  });

  const [targetId, setTargetId] = useState<string>("");
  const [lockToRelease, setLockToRelease] = useState<number>(0);
  const [isReleasing, setIsReleasing] = useState(false);

  const { control, formState, handleSubmit } = useForm<{ search: string }>({});

  const { roles } = useRolesStore();

  const { isOpen, onClose, onOpen } = useDisclosure();

  const {
    data: lockers,
    isInitialLoading: isInitLoading,
    isLoading,
    isFetching,
  } = useGetLockers(
    new GetLocksIdRequest({
      securityId: id,
      targetId,
      start: 0,
      end: 10,
    }),
    {
      cacheTime: 0,
      staleTime: 0,
      enabled: !!targetId,
      refetchOnWindowFocus: false,
      select: (data) => {
        return data.sort((a, b) => {
          return Number(a.expirationDate) - Number(b.expirationDate);
        });
      },
    },
  );

  const { mutate: releaseMutation } = useRelease();

  const hasLockerRole = roles.includes(SecurityRole._LOCKER_ROLE);

  const onSubmit = ({ search }: { search: string }) => {
    setTargetId(search);
  };

  const handleRelease = () => {
    setIsReleasing(true);

    const request = new ReleaseRequest({
      securityId: id,
      targetId,
      lockId: lockToRelease,
    });

    releaseMutation(request, {
      onSettled: async () => {
        setIsReleasing(false);
        onClose();
        setLockToRelease(0);
      },
      onSuccess: () => {
        queryClient.setQueryData(
          [GET_LOCKERS(id, targetId)],
          (oldData: LockViewModel[] | undefined) => {
            return oldData?.filter((lock) => lock.id !== lockToRelease);
          },
        );
        queryClient.invalidateQueries({
          queryKey: [GET_LOCKERS(id, targetId)],
        });
      },
    });
  };

  return (
    <VStack gap={4} h="full">
      <PopUp
        id="confirmReleasePopUp"
        isOpen={isOpen}
        onClose={onClose}
        icon={<PhosphorIcon as={WarningCircle} size="md" />}
        title={tRelease("confirmPopUp.title")}
        description={tRelease("confirmPopUp.description")}
        confirmText={tRelease("confirmPopUp.confirmText")}
        onConfirm={() => {
          onClose();
          handleRelease();
        }}
        onCancel={() => {
          onClose();
          setLockToRelease(0);
        }}
        cancelText={tRelease("confirmPopUp.cancelText")}
        confirmButtonProps={{ status: "danger" }}
      />
      <HStack w="full" justifyContent="flex-end" gap={4}>
        {hasLockerRole && (
          <Button
            data-testid="locker-button"
            as={RouterLink}
            to={RouterManager.getUrl(RouteName.DigitalSecurityLock, {
              params: { id },
            })}
            variant="secondary"
          >
            {tButtons("lock")}
          </Button>
        )}
      </HStack>
      <Stack w="full" h="full" layerStyle="container">
        <Center w="full" h="full" bg="neutral.dark.600">
          <VStack align="flex-start" p={6} gap={4}>
            <Heading textStyle="HeadingMediumLG">{t("search.title")}</Heading>
            <Text textStyle="BodyRegularMD">{t("search.description")}</Text>
            <HStack
              w="440px"
              gap={6}
              mt={3}
              as="form"
              onSubmit={handleSubmit(onSubmit)}
              alignItems="flex-start"
            >
              <SearchInputController
                id="search"
                placeholder={t("search.search")}
                control={control}
                onSearch={() => {}}
                size="sm"
                rules={{
                  required,
                  validate: { isValidHederaId: isValidHederaId },
                }}
              />
              <Button
                data-testid="search-button"
                size="sm"
                type="submit"
                isLoading={isInitLoading || isFetching}
                isDisabled={!formState.isValid || isInitLoading}
              >
                <Text textStyle="ElementsMediumSM" px={4}>
                  {t("search.search")}
                </Text>
              </Button>
            </HStack>
          </VStack>
        </Center>
      </Stack>
      <Stack w="full" h="full" gap={10}>
        {lockers && lockers.length && !isLoading && (
          <VStack align="flex-start" gap={4}>
            {lockers.map((lock, index) => {
              const canBeReleased =
                Number(lock.expirationDate) * 1000 < Date.now();

              return (
                <Center key={index} w="full" h="full" layerStyle="container">
                  <VStack w="440px" align="flex-end" py={4} gap={4}>
                    <DefinitionList
                      style={{ paddingBottom: 0 }}
                      items={[
                        { title: t("list.lockId"), description: lock.id },
                        { title: t("list.amount"), description: lock.amount },
                        {
                          title: t("list.expirationDate"),
                          description: formatDate(
                            Number(lock.expirationDate) * 1000,
                            DATE_TIME_FORMAT,
                          ),
                        },
                      ]}
                    />
                    {canBeReleased && hasLockerRole && (
                      <Button
                        size={"sm"}
                        disabled={isReleasing && lockToRelease === lock.id}
                        isLoading={isReleasing && lockToRelease === lock.id}
                        onClick={() => {
                          onOpen();
                          setLockToRelease(lock.id);
                        }}
                      >
                        {t("list.release")}
                      </Button>
                    )}
                  </VStack>
                </Center>
              );
            })}
          </VStack>
        )}
        {lockers && lockers.length === 0 && !isLoading && (
          <Center>
            <Text>{t("list.noLocks")}</Text>
          </Center>
        )}
      </Stack>
    </VStack>
  );
};
