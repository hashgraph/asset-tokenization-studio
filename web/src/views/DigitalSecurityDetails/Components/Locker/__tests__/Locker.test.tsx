import { fireEvent, waitFor } from "@testing-library/react";
import { render } from "../../../../../test-utils";
import { Locker } from "../Locker";

describe(`${Locker.name}`, () => {
  test("should render correctly", () => {
    const component = render(<Locker />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("should render the search button disabled if the targetId is empty", () => {
    const component = render(<Locker />);

    const searchButton = component.getByTestId("search-button");

    expect(searchButton).toBeDisabled();
  });

  test("should render the search button enabled if the targetId is not empty", async () => {
    const component = render(<Locker />);

    const searchButton = component.getByTestId("search-button");

    expect(searchButton).toBeDisabled();

    fireEvent.change(component.getByTestId("search"), {
      target: { value: "0.0.12345" },
    });

    await waitFor(() => {
      expect(searchButton).toBeEnabled();
    });
  });
});
