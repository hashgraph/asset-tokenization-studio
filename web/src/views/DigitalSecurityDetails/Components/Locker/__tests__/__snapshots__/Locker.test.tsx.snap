// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locker should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-iv2p4a"
  >
    <div
      class="chakra-stack css-1xa9bwv"
    />
    <div
      class="chakra-stack css-r8fju2"
    >
      <div
        class="css-b2xmzl"
      >
        <div
          class="chakra-stack css-1c05t7c"
        >
          <h2
            class="css-1mk5tlk"
          >
            Display Locks
          </h2>
          <p
            class="css-j9h4p6"
          >
            Add the ID account to preview its locks
          </p>
          <form
            class="chakra-stack css-1y1dtud"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="css-1fey8w1"
                for="search"
              >
                <div
                  class="chakra-input__group css-pwmwmr"
                  data-group="true"
                >
                  <input
                    class="chakra-input css-1ymw93l"
                    data-testid="search"
                    id="search"
                    name="search"
                    placeholder="Search"
                    value=""
                  />
                  <div
                    class="chakra-input__right-element css-1akgnmh"
                  >
                    <button
                      aria-label="search"
                      class="chakra-button css-kdxsnu"
                      data-testid="search-icon-button"
                      disabled=""
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="chakra-icon css-kpbtv0"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M229.66,218.34l-50.07-50.06a88.11,88.11,0,1,0-11.31,11.31l50.06,50.07a8,8,0,0,0,11.32-11.32ZM40,112a72,72,0,1,1,72,72A72.08,72.08,0,0,1,40,112Z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </label>
            </div>
            <button
              class="chakra-button css-gobc5v"
              data-testid="search-button"
              disabled=""
              type="submit"
            >
              <p
                class="css-15hpv2u"
              >
                Search
              </p>
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-fopn24"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
